x/datadeal/keeper/deal.go

```diff
@@ -544,45 +550,82 @@ func (k Keeper) DeactivateDeal(ctx sdk.Context, msg *types.MsgDeactivateDeal) er
 		return sdkerrors.Wrapf(types.ErrDealDeactivate, "deal's status is not 'ACTIVE'")
 	}
 
-	deal.Status = types.DEAL_STATUS_INACTIVE
+	deal.Status = types.DEAL_STATUS_DEACTIVATING
+
+	err = k.SetDeal(ctx, deal)
+	if err != nil {
+		return sdkerrors.Wrapf(types.ErrDealDeactivate, err.Error())
+	}
+
+	params := k.oracleKeeper.GetParams(ctx)
+	VotingPeriod := params.VoteParams.VotingPeriod
+
+	deactivationHeight := ctx.BlockHeader().Height + DealDeactivationParam*int64(VotingPeriod/BlockPeriod) +1
+
+	k.AddDealQueue(ctx, deal.Id, deactivationHeight)
+
+	return nil
+}
+
+func (k Keeper) DeactivateDeal(ctx sdk.Context, dealID uint64) error {
+	deal, err := k.GetDeal(ctx, dealID)
+	if err != nil {
+		return sdkerrors.Wrapf(types.ErrDealDeactivate, err.Error())
+	}
+
+	if deal.Status != types.DEAL_STATUS_DEACTIVATING {
+		return sdkerrors.Wrapf(types.ErrDealDeactivate, "deal's status is not 'DEACTIVATING'")
+	}
+
+	// change deal status to DEAL_STATUS_DEACTIVATED
+	deal.Status = types.DEAL_STATUS_DEACTIVATED
 
 	err = k.SetDeal(ctx, deal)
 	if err != nil {
 		return sdkerrors.Wrapf(types.ErrDealDeactivate, err.Error())
 	}
 
+	// refund remaining budget to buyer
 	dealAcc, err := sdk.AccAddressFromBech32(deal.Address)
 	if err != nil {
 		return sdkerrors.Wrapf(types.ErrDealDeactivate, err.Error())
 	}
 
-	requesterAcc, err := sdk.AccAddressFromBech32(msg.RequesterAddress)
+	buyerAcc, err := sdk.AccAddressFromBech32(deal.BuyerAddress)
 	if err != nil {
 		return sdkerrors.Wrapf(types.ErrDealDeactivate, err.Error())
 	}
 
 	dealBalance := k.bankKeeper.GetBalance(ctx, dealAcc, assets.MicroMedDenom)
 
-	err = k.bankKeeper.SendCoins(ctx, dealAcc, requesterAcc, sdk.Coins{dealBalance})
+	err = k.bankKeeper.SendCoins(ctx, dealAcc, buyerAcc, sdk.Coins{dealBalance})
 	if err != nil {
 		return sdkerrors.Wrapf(types.ErrDealDeactivate, err.Error())
 	}
 
-	//TODO: Implement Remove the DataVerification/DeliveryVote Queue after PR #449 merged
-	//https://github.com/medibloc/panacea-core/pull/449
+	//Todo:Remove DataVerification/DeliveryVote Queue
 
 	return nil
 }
 
 func (k Keeper) ReRequestDataDeliveryVote(ctx sdk.Context, msg *types.MsgReRequestDataDeliveryVote) error {
 
+	deal, err := k.GetDeal(ctx, msg.DealId)
+	if err != nil {
+		return sdkerrors.Wrapf(types.ErrReRequestDataDeliveryVote, err.Error())
+	}
+
+	if deal.Status != types.DEAL_STATUS_ACTIVE && deal.Status != types.DEAL_STATUS_COMPLETED {
+		return sdkerrors.Wrapf(types.ErrReRequestDataDeliveryVote, "not in deal state to make a request")
+	}
```

Comment:
We ensure that we receive all requested sales data.
So I think that we don't need check deal status.

Reply:
AFAIK, we need to check the status because there is no budget for data delivery in deactivated deal.

Reply:
I forgot this spec!
Please ignore my comment.
Thank you.